download/tests/read-proxy-env.rs

```diff
@@ -28,3 +31,37 @@ fn read_basic_proxy_params() {
         Some(("proxy.example.com".to_string(), 8080))
     );
 }
+
+// Tests to verify if socks5 feature is available and being used
+#[test]
+fn socks_proxy_request() {
+    use std::sync::atomic::{AtomicUsize, Ordering};
+    static CALL_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    scrub_env();
+    set_var("all_proxy", "socks5://192.0.2.0:1080"); //TEST-NET IP address
+
+    let env_proxy = |url: &Url| {
+        let url = for_url(&url).to_url();
+        CALL_COUNT.fetch_add(1, Ordering::SeqCst);
+        url
+    };
+
+    let url = Url::parse("http://example.org").unwrap();
+    let client = Client::builder()
+        .proxy(Proxy::custom(env_proxy))
+        .timeout(Duration::from_secs(1))
+        .build()
+        .unwrap();
+
+    let res = client.get(url.as_str()).send();
+    println!("Called {}", CALL_COUNT.load(Ordering::SeqCst));
+    if let Err(e) = res {
+        let s = e.source().unwrap();
+        // Only 2 calls performed in case of lack socks5 support
```

Comment:
This number feels very fragile to me.  Could we instead create a socket, bind and listen, and then set the proxy to that socket and check that it tries to connect to it?

Reply:
I've modified test to actually check if proxy connects to the socket. But I noticed that sometimes tests can overlap and mess around with env variables. I see 3 options to fix this:
1. Use `serial_test` [crate](https://docs.rs/serial_test/0.4.0/serial_test/) to make those 2 tests simultaneous
2. Use Mutex for the same purpose
3. Or just make tests use different variables
Not sure though which is the best approach in this case

Reply:
Individual integration tests ought to be being run in isolation; though on Windows environment changes are moderately awkward.  We have some other tests for environment in the download crate and they use a mutex to serialise them, perhaps move the test to be like those?

Reply:
I have made a change which I believe will work.